### PR TITLE
Share Extension: Allows multiple media items to be uploaded

### LINKED ISSUE #8179: Share Extension: Allow multiple media items to be shared
Currently, we only can share a single image. Let's enable the sharing of multiple media items within the existing `SLComposeServiceViewController` based UI. Things to be done:

- [x] Save selected media files from host app to shared container
- [x] Create a small, separated CoreData database in the shared container to store pending operations
- [x] Kickoff background upload in share extension (started in https://github.com/wordpress-mobile/WordPress-iOS/pull/8008)
- [x] Extend `ShareExtensionSessionManager` so WPiOS can work through the remaining pending operations (staged in the shared container db) from ☝️ 
- [x] Discontinue use of existing endpoint to create posts (to prevent galleries in post)

### PR DESCRIPTION
**Pull request by bummytime**
## Discussion
This PR changes the way we are sharing media from the existing `SLComposeServiceViewController`-based Share VC. It is core to allowing us to leverage the new app extension designs (e.g. #8137, #8138). There are a few things to note here:

**1.** The biggest change is that there is a new, separate core data stack (`SharedCoreDataStack.swift`) that persists a single managed object (`UploadOperation.swift`) to the shared container. This CD instance the fundamental communication mechanism between app extensions and WPiOS. Each "UploadOp" represents individual media items as well as the post itself. For example, if you share a post with two attached images, there will be 3 UploadOps persisted to the store — 2 records for the images, and 1 record for the post. As each upload is processed, the corresponding upload op is updated with a status, (Pending, In Progress, Complete, or Error). 

**2.** You will undoubtedly notice that we have to slightly duplicate logic between `ShareViewController.swift` and `ShareExtensionSessionManager.swift` ... specifically the completion handlers for `NSURLSession` background sessions. Unfortunately there is no way around this. It is non-deterministic as to whether the share extension or WPiOS will get the call (9 times out of 10, WPiOS does like you would expect).

### General Workflow
1. User selects 1 or more images to share in the host app.
2. `ShareExtractor` parses the context extension and extracts the images and/or text
3. User is presented with `ShareViewController`
4. If user selects "Post" on dialog:
    - The selected media is encoded and saved into the shared container
    - Upload ops are created in CD for each piece of media and the post
    - A NSURLSession background session is kicked off to upload the media
5. Server response is executed:
    - **Scenario 1:** The media upload happens quickly and the NSURLSession background session completion handler runs inside the share extension (`ShareViewController`).
    - **Scenario 2 :** The NSURLSession background session completion handler runs inside WPiOS (`ShareExtensionSessionManager`).
6. Based on the response in step 5, the corresponding upload ops are updated in the shared CD instance. 
7. The local files media files in the shared container are removed IF the server response was 👍 (not deleted if it wasn't)
8. Using the post upload op, a new post is uploaded to the server containing the remote media.
9. New post response is received and saved to the CD instance.

### Things to note:
* Because app extensions are notoriously problematic, we are taking a conservative approach and only allowing for a **max of 2 images** to be shared. We can increase this limit in a subsequent PR when we release the new custom UI _and_ see if there are any crashers reported when 9.1 is pushed.

* Keep in mind that `ShareViewController`'s days are numbered, but we will transfer a lot of the logic into the new UI classes.

* There is no retry (yet)

Fixes #8179
Ref: #8265 and #8263 

### App extension fun
A few things to think about when you are testing:

* Xcode’s debugger prevents the system from suspending the process being debugged.  So, if you run your extension from Xcode, or you attach to its process some time after launch, the process will continue to execute in situations where the system would otherwise have suspended it.  
* Be sure to test on a device, as well as the simulator. The simulator is much less likely to terminate an extension.
* Try to test on a variety of devices iPhone 5s, iPhone X, etc.
* If the user terminates the app by force-quiting from the app switcher, the system will cancel all of the session’s background transfers, and won’t attempt to relaunch the app.

### What to test
The main thing we want to test are all of the sharing flows with media. Example flow:

1. Build and run the app on a device (NOT connected to XCode)
2. Open the photos app
3. Select 1 or 2 photos and share using the debug build
4. Hopefully see a post on your site within 10 Seconds to 3 minutes depending on connection.

Try to test different scenarios/edge cases:
* Kill the network connectivity while the background sesson is running and then turn it on again (it should eventually resume but may take a few minutes).
* Multitasking on iPad
* Share images from other apps
* Share from the host app and then quickly open WPiOS (and leave it open)
* Try a really slow connection
* Try sharing an image and then after the share UI is dismissed, quickly share another image (to create another bg session)

## Thank you!

@SergioEstevao, @frosty, @kwonye I would appreciate it if you could take a look at this! Please let me know if you have any questions!

### CODE REVIEW

**Comment by SergioEstevao** [comment]
@bummytime Just give it a first spin and its working very well, but there is small detail that you may want to fix. So when you insert the image on the post you are just writing:

``` ```

but you need to write this:

``````

This is to match the default behaviour in Calypso and in order for the images to be detected as possible featured images.

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 1]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+            if let error = error as NSError? {
+                errorString += ": \(error.localizedDescription)"
+            }
+            self.logError(errorString: errorString, uploadOpObjectIDs: postUploadOpID)
+            self.cleanupSessionAndTerminate()
+        })
+    }
 }
 
-extension ShareExtensionSessionManager: URLSessionTaskDelegate {
```
Comment (SergioEstevao) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: I feel that all these methods related to the core data operations should move to a separate class something like ShareExtensionStore.

File: WordPress/WordPressShareExtension/ShareViewController.swift [hunk 2]
```diff
@@ -243,66 +277,165 @@ private extension ShareViewController {
...
+            fatalError("Failed to fetch UploadOperations: \(error)")
+        }
+
+        return mediaUploadOps
+    }
+}
+
+// MARK: - ShareViewController Extension: Backend Interaction
+
 private extension ShareViewController {
```
Comment (SergioEstevao) [inline WordPress/WordPressShareExtension/ShareViewController.swift]: These methods on this extension could also move to the ShareExtensionStore class.

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 3]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+            logError(errorString: "Error creating post: site ID was missing.", uploadOpObjectIDs: postUploadOpID)
+            return
+        }
+
+        let api = WordPressComRestApi(oAuthToken: oauth2Token,
+                                      userAgent: nil,
+                                      backgroundUploads: false,
+                                      backgroundSessionIdentifier: backgroundSessionIdentifier,
+                                      sharedContainerIdentifier: WPAppGroupName)
+        let remote = PostServiceRemoteREST.init(wordPressComRestApi: api, siteID: NSNumber(value: postUploadOp.siteID))
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: I think you can remove `.init` here?

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 4]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+        postUploadOp.currentStatus = .InProgress
+        coreDataStack.saveContext()
+
+        remote.createPost(postUploadOp.remotePost, success: { post in
+            if let post = post {
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: Could switch this to a `guard` just to reduce the indentation slightly.

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 5]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+                if let postID = post.postID {
+                    DDLogInfo("Post \(post.postID.stringValue) sucessfully uploaded to site \(post.siteID.stringValue)")
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: Nitpick, typo: `sucessfully`

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 6]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+    /// file name. If a parameter is nil, that specific param will not be updated.
+    ///
+    /// Note: We are searching for the upload op using a filename because a given task ID can have
+    /// multiple files associated with it.
+    ///
+    /// - Parameters:
+    ///   - fileName: the fileName from a URLSessionTask
+    ///   - remoteMediaID: remote media ID. Can be nil.
+    ///   - remoteURL: remote media URL string. Can be nil.
+    func updateMediaOperationForFileName(_ fileName: String, remoteMediaID: Int64?, remoteURL: String?) {
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: As above, these methods could probably lose the `filename` from the first part of the method name - `updateMediaOperation(for fileName: String, remoteMediaID...)`?

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift [hunk 7]
```diff
@@ -51,23 +56,315 @@ import Foundation
...
+        guard let uploadMediaOp = fetchUploadOpForFileName(fileName) else {
+            DDLogError("Error loading UploadOperation Object with File Name: \(fileName)")
+            return
+        }
+
+        if let remoteID = remoteMediaID {
```
Comment (frosty) [inline WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: Do we need to unwrap these before we assign them? The property in the core data is optional, so could we update the Swift property declaration to be optional too?

One other note – When doing unwrapping like this, you can use the same variable name: 

`if let remoteMediaID = remoteMediaID {` for example :)

File: WordPress/WordPressShareExtension/UploadOperation.swift [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+// MARK: - UploadOperation Types
+
+extension UploadOperation {
+    /// Status types for a given upload operation
+    ///
+    enum UploadStatus: Int {
+        /// Upload has been queued, but not started
+        ///
+        case Pending
```
Comment (frosty) [inline WordPress/WordPressShareExtension/UploadOperation.swift]: To match Swift API guidelines, these cases should ideally be all lowercase :)

**Comment by bummytime** [comment]
Thanks for the reviews @SergioEstevao and @frosty. Changes incoming!

File: WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift (hunk 1 above)
Comment (bummytime) [reply WordPress/WordPressShareExtension/ShareExtensionSessionManager.swift]: Seems like a good plan to me. 👍

File: WordPress/WordPressShareExtension/Extensions.xcdatamodeld/Extensions.xcdatamodel/contents [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<model type="com.apple.IDECoreDataModeler.DataModel" documentVersion="1.0" lastSavedToolsVersion="13533" systemVersion="17B48" minimumToolsVersion="Automatic" sourceLanguage="Swift" userDefinedModelVersionIdentifier="">
+    <entity name="UploadOperation" representedClassName="UploadOperation" syncable="YES">
```
[the remark this answers is not in the document]
Comment (bummytime) [reply WordPress/WordPressShareExtension/Extensions.xcdatamodeld/Extensions.xcdatamodel/contents]: Yeah, I mulled on this as well. I was trying to keep the data model "simple" by having a single table, but I think you are right. Separate entities makes sense here...I will update.

File: WordPress/WordPressShareExtension/ShareViewController.swift [hunk 10]
```diff
@@ -243,66 +277,165 @@ private extension ShareViewController {
...
         self.mediaView = mediaView
         self.reloadConfigurationItems()
     }
 }
 
-/// ShareViewController Extension: Backend Interaction
-///
+// MARK: - ShareViewController Extension: Persistence
+
+private extension ShareViewController {
```
[the remark this answers is not in the document]
Comment (bummytime) [reply WordPress/WordPressShareExtension/ShareViewController.swift]: If it's ok with you @frosty , I think I am going to move this logic out of the VC since we have a new one replacing it very soon (and it could be shared with WPiOS as well).

File: WordPress/WordPressShareExtension/ShareViewController.swift [hunk 11]
```diff
@@ -62,6 +63,17 @@ class ShareViewController: SLComposeServiceViewController {
         return identifier
     }()
 
+    /// Unique identifier a group of upload operations
+    ///
+    fileprivate lazy var groupIdentifier: String = {
+        let groupIdentifier = UUID().uuidString
+        return groupIdentifier
```
[the remark this answers is not in the document]
Comment (bummytime) [reply WordPress/WordPressShareExtension/ShareViewController.swift]: @frosty For now, this needs to remain because of the check that the WPiOS delegate is doing in `handleEventsForBackgroundURLSession`

File: WordPress/WordPressShareExtension/UploadOperation.swift [hunk 12]
```diff
@@ -0,0 +1,192 @@
+import Foundation
+import CoreData
+import WordPressKit
+
+@objc(UploadOperation)
```
[the remark this answers is not in the document]
Comment (bummytime) [reply WordPress/WordPressShareExtension/UploadOperation.swift]: @frosty I did attempt to remove this, however CD had a bunch of fits so I left them in for now.

**Comment by bummytime** [comment]
@SergioEstevao and @kwonye This PR is ready for another look. I made almost all of the changes suggested above ☝️ including splitting the `UploadOp` MO into two separate MO's (`MediaUploadOperation` and `PostUploadOperation`) with a common parent for shared props.

/cc @elibud and @diegoreymendez: I would like land this in time for the 9.1 freeze 🤞

File: WordPress/WordPressShareExtension/ShareViewController.swift [hunk 13]
```diff
@@ -243,66 +277,165 @@ private extension ShareViewController {
...
+                errorString += ": \(error.localizedDescription)"
             }
+            DDLogError(errorString)
+            self.updateStatus(.Error, forUploadOpWithObjectID: uploadOpObjectID)
             requestEnqueued()
         })
     }
 
-    func uploadPostWithMedia(subject: String, body: String, status: String, siteID: Int, attachedImageData: Data?, requestEnqueued: @escaping () -> ()) {
+    func uploadPostWithMedia(subject: String, body: String, status: String, siteID: Int, attachedImageData: [Data?]?, requestEnqueued: @escaping () -> ()) {
```
[the remark this answers is not in the document]
Comment (bummytime) [reply WordPress/WordPressShareExtension/ShareViewController.swift]: @frosty Since `ShareViewController' is being retired, we can tackle this in a subsequent PR. :-)

**Comment by diegoreymendez** [comment]
Just a reminder that the freeze for 9.1 is scheduled for today.

Let's make sure this PR is merged as soon as possible to assist @elibud with the process.

Ping @bummytime, @SergioEstevao, @frosty.

**Comment by bummytime** [comment]
Thanks @kwonye. We do [retrieve](https://github.com/wordpress-mobile/WordPress-iOS/blob/bd4b93641903b5080ea5ba4a9640d8c8d5e62b8f/WordPress/WordPressShareExtension/ShareViewController.swift#L162) the maximum image dimension from `NSUserDefaults` [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/bd4b93641903b5080ea5ba4a9640d8c8d5e62b8f/WordPress/Classes/Services/ShareExtensionService.swift#L158-L164). It appears to be working as expected, but if you were seeing a specific issue, let me know and we can file an issue.

**Comment by kwonye** [comment]
> We do retrieve the maximum image dimension from NSUserDefaults here. It appears to be working as expected

Ok, I didn't catch that. That's perfect then!

**Comment by kwonye** [comment]
I'm removing @frosty's block from this PR as he's gone for the next two weeks.
